### PR TITLE
Cert submitter optimization

### PR DESCRIPTION
**Pull request by VolodymyrBabaiev**

### CODE REVIEW

**Comment by VolodymyrBabaiev** [comment]
Mainchain part for resolving the issue from sidechain https://github.com/HorizenOfficial/Sidechains-SDK/issues/256
Also contain the fixes for https://github.com/HorizenOfficial/zend_oo/issues/114 (doesn't have tests).

**Review by vovanloc0798 — APPROVED** [review]

File: src/zen/websocket_server.cpp [hunk 1]
```diff
@@ -488,6 +507,66 @@ class WsHandler
...
+                const uint256& topQualCertHash = mempool.mapSidechains.at(scId).GetTopQualityCert()->second;
+                const CScCertificate& topQualCert = mempool.mapCertificate.at(topQualCertHash).GetCertificate();
+                CDataStream ssBlock(SER_NETWORK, PROTOCOL_VERSION);
+                ssBlock << topQualCert;
+                std::string certHex = HexStr(ssBlock.begin(), ssBlock.end());
+     
+                mempoolTopQualityCert.push_back(Pair("quality", topQualCert.quality));
+                mempoolTopQualityCert.push_back(Pair("certHash", topQualCertHash.GetHex()));
+                mempoolTopQualityCert.push_back(Pair("rawCertificateHex", certHex));
+                mempoolTopQualityCert.push_back(Pair("fee", topQualCert.mainchainBackwardTransferRequestScFee));
```
Comment (i-Alex) [inline src/zen/websocket_server.cpp:538]: Mempool Cert fee is different. Fees are the coins amount diff between inputs and outputs.
There is an info about cert fees in the mempool.

File: qa/rpc-tests/ws_messages.py [hunk 2]
```diff
@@ -372,5 +375,71 @@ def run_test(self):
...
+        assert_equal(cert_epoch_0, mempool_cert_['certHash'])
+        assert_equal({}, chain_cert_)
+
+        self.nodes[0].generate(1)
+        assert_equal(False, cert_epoch_0 in self.nodes[0].getrawmempool())
+
+        mempool_cert_, chain_cert_ = self.nodes[0].ws_get_top_quality_certificates(scid2)
+        assert_equal(cert1_quality, chain_cert_['quality'])
+        assert_equal(cert_epoch_0, chain_cert_['certHash'])
+        assert_equal({}, mempool_cert_)
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply qa/rpc-tests/ws_messages.py:445]: Done.

File: src/zen/websocket_server.cpp [hunk 3]
```diff
@@ -488,6 +507,66 @@ class WsHandler
...
+            }
+            
+            CSidechain sidechainInfo;
+
+            if (view.GetSidechain(scId, sidechainInfo)) {
+                const uint256& topQualCertHash = sidechainInfo.lastTopQualityCertHash;
+                const int topQualityCertQuality = sidechainInfo.lastTopQualityCertQuality;
+                CScCertificate topQualCert;
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply src/zen/websocket_server.cpp:547]: Done.

**Review by i-Alex — APPROVED** [review]

**Review by i-Alex — APPROVED** [review]

File: src/zen/websocket_server.cpp [hunk 4]
```diff
@@ -227,6 +229,24 @@ class WsHandler
...
+    
+    void sendTopQualityCertificates(const UniValue& mempoolCert, const UniValue& chainCert,
+                                    WsEvent::WsMsgType msgType, std::string clientRequestId = "")
+    {
+        // Send a message to the client:  type = eventType
+        WsEvent* wse = new WsEvent(msgType);
+        LogPrint("ws", "%s():%d - allocated %p\n", __func__, __LINE__, wse);
+        UniValue rspPayload(UniValue::VOBJ);
+        
+        rspPayload.push_back(Pair("mempool_top_quality_cert", mempoolCert));
```
Comment (i-Alex) [inline src/zen/websocket_server.cpp:241]: let's use CamelCase names as in other places:
`mempool_top_quality_cert` -> `mempoolTopQualityCert`
`chain_top_quality_cert` -> `chainTopQualityCert`

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: src/zen/websocket_server.cpp [hunk 5]
```diff
@@ -488,6 +508,68 @@ class WsHandler
...
+                if (!topQualCertHash.IsNull()) {
+                    uint256 blockHash;
+                    if (GetCertificate(topQualCertHash, topQualCert, blockHash)) {
+                        CDataStream ssBlock(SER_NETWORK, PROTOCOL_VERSION);
+                        ssBlock << topQualCert;
+                        std::string certHex = HexStr(ssBlock.begin(), ssBlock.end());
+
+                        chainTopQualityCert.push_back(Pair("quality", topQualityCertQuality));
+                        chainTopQualityCert.push_back(Pair("certHash", topQualCertHash.GetHex()));
+                        chainTopQualityCert.push_back(Pair("rawCertificateHex", certHex));
```
Comment (alsala) [inline src/zen/websocket_server.cpp:559]: Just for symmetry with mempool, should we need the fee in the json obj we could use the view obj for getting it from the cert, for instance:
   certFee = topQualCert.GetFeeAmount(view.GetValueIn(topQualCert));

**Review by alsala — CHANGES_REQUESTED** [review]

File: src/zen/websocket_server.cpp [hunk 6]
```diff
@@ -488,6 +508,68 @@ class WsHandler
...
+                mempoolTopQualityCert.push_back(Pair("rawCertificateHex", certHex));
+                mempoolTopQualityCert.push_back(Pair("fee", FormatMoney(certFee)));
+            }
+
+            CSidechain sidechainInfo;
+
+            if (view.GetSidechain(scId, sidechainInfo)) {
+                const uint256& topQualCertHash = sidechainInfo.lastTopQualityCertHash;
+                const int topQualityCertQuality = sidechainInfo.lastTopQualityCertQuality;
+                CScCertificate topQualCert;
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply src/zen/websocket_server.cpp:547]: Done.

**Review by alsala — APPROVED** [review]

**Review by i-Alex — APPROVED** [review]

File: src/zen/websocket_server.cpp [hunk 7]
```diff
@@ -488,6 +508,65 @@ class WsHandler
...
+                uint256 blockHash;
+
+                if (GetCertificate(sidechainInfo.lastTopQualityCertHash, topQualCert, blockHash)) {
+                    CDataStream ssCert(SER_NETWORK, PROTOCOL_VERSION);
+                    ssCert << topQualCert;
+                    std::string certHex = HexStr(ssCert.begin(), ssCert.end());
+
+                    chainTopQualityCert.push_back(Pair("quality", topQualityCertQuality));
```
Comment (albertog78) [inline src/zen/websocket_server.cpp:555]: return also as a separate field the certificate referenced epoch (we have to be sure on sidechain side that the last received certificate doesn't belong to the previous epoch)
Do the same for mempool cert

**Review by albertog78 — CHANGES_REQUESTED** [review]

**Review by i-Alex — APPROVED** [review]

**Review by albertog78 — APPROVED** [review]

**Pull request merged**